compiler/rustc_parse/src/parser/path.rs

```diff
@@ -934,6 +943,10 @@ impl<'a> Parser<'a> {
             }
         } else if self.token.is_keyword(kw::Const) {
             return self.recover_const_param_declaration(ty_generics);
+        } else if let Some(attr_span) = attr_span {
+            // Handle case where attribute is present but no valid generic argument follows.
+            let guar = self.dcx().emit_err(AttributeOnEmptyGeneric { span: attr_span });
+            return Ok(Some(GenericArg::Type(self.mk_ty(attr_span, TyKind::Err(guar)))));
```

Comment:
Instead of returning a poisoned *type* (which is undesirable as discussed), simply bail out.

```suggestion
            let diag = self.dcx().create_err(AttributeOnEmptyGeneric { span: attr_span });
            return Err(diag);
```

Reply:
> return Err(diag) stops parsing after the first error, which is slightly overkill because making 5 tests for each case is fundamental wrong

`return Err(diag);` is the least hacky and the simplest solution. We can always iterate from here via follow-up PRs. Three test cases are basically identical:

* `let _: Baz<#[cfg(any())]> = todo!();`,
* `let _: Foo<#[cfg(not(wrong))]> = todo!();` and
* `let _: Bar<#[cfg(any())]> = todo!();`.

They should just be combined into one. Lastly, you can move the resulting unified "empty gen arg" case to the end of the file, so rustc still emits the other ones before it bails out.

To explain myself a bit more, the "empty gen args" case is more unlikely to happen in practice than all the other cases, so it's in my eyes absolutely justifiable that its logic is more conservative. We can rejoice, free of any headaches, since we know there won't be any invalid or weird consequential diagnostics and the impl is simple.

Reply:
So completly kills parser?